packages/localize/src/program-analysis.ts

```diff
@@ -86,112 +88,250 @@ function extractMsg(
+        file,
+        node,
+        `Expected second argument to msg() to be an object`
+      ),
     };
   }
 
-  if (isLitTemplate(contentsArg)) {
-    if (ts.isNoSubstitutionTemplateLiteral(contentsArg.template)) {
-      // E.g. msg('foo', html`bar <b>baz</b>`)
+  let id: string | undefined = undefined;
+  let args: ts.NodeArray<ts.Expression> | undefined = undefined;
+
+  for (const property of node.properties) {
+    if (!ts.isPropertyAssignment(property)) {
       return {
-        name,
-        contents: combineAdjacentPlaceholders(
-          replaceHtmlWithPlaceholders(contentsArg.template.text)
+        error: createDiagnostic(
+          file,
+          property,
+          `Options shorthand object assignment is not supported`
+        ),
+      };
+    }
+
+    let name;
+    if (ts.isIdentifier(property.name)) {
+      name = property.name.escapedText;
+    } else if (ts.isStringLiteral(property.name)) {
+      name = property.name.text;
+    } else {
+      return {
+        error: createDiagnostic(
+          file,
+          property.name,
+          `Options shorthand object assignment must be identifier or string literal`
+        ),
+      };
+    }
+
+    if (name === 'id') {
+      if (
+        (!ts.isStringLiteral(property.initializer) &&
+          !ts.isNoSubstitutionTemplateLiteral(property.initializer)) ||
+        property.initializer.text.trim() === ''
+      ) {
+        return {
+          error: createDiagnostic(
+            file,
+            property.initializer,
+            `Options id property must be a non-empty string literal`
+          ),
+        };
+      }
+      id = property.initializer.text;
+    } else if (name === 'args') {
+      if (!ts.isArrayLiteralExpression(property.initializer)) {
+        return {
+          error: createDiagnostic(
+            file,
+            property.initializer,
+            `Options args property must be an array literal`
+          ),
+        };
+      }
+      args = property.initializer.elements;
+    } else {
+      return {
+        error: createDiagnostic(
+          file,
+          property,
+          `Options shorthand object property must be "id" or "args"`
         ),
-        node,
-        file,
-        descStack: descStack.map((desc) => desc.text),
-        isLitTemplate: true,
       };
     }
-    // E.g. msg('foo', html`bar ${baz}`)
-    return createDiagnostic(
-      file,
-      contentsArg,
-      `To use a variable, pass an arrow function.`
-    );
   }
 
-  if (ts.isTemplateExpression(contentsArg)) {
-    // E.g. msg('foo', `bar ${baz}`)
-    return createDiagnostic(
-      file,
-      contentsArg,
-      `To use a variable, pass an arrow function.`
-    );
+  return {result: {id, args}};
+}
+
+/**
+ * Analyze the template argument to a msg call.
+ */
+export function extractTemplate(
+  templateArg: ts.Node,
+  file: ts.SourceFile
+): ResultOrError<
+  Pick<ProgramMessage, 'contents' | 'params' | 'isLitTemplate'> & {
```

Comment:
nit: I think it'd be nice to give this type a name

Reply:
Done (`ExtractedTemplate`).